File: src/main/java/com/syfe/personalfinance/dto/TransactionDto.java
package com.syfe.personalfinance.dto;

import com.syfe.personalfinance.enums.CategoryType;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PastOrPresent;
import jakarta.validation.constraints.Positive;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

public class TransactionDto {

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class CreateTransactionRequest {
        @NotNull(message = "Transaction amount is required")
        @Positive(message = "Amount must be a positive number")
        private BigDecimal amount;

        @NotNull(message = "Transaction date is required")
        @PastOrPresent(message = "Transaction date cannot be in the future")
        private LocalDate date;

        @NotBlank(message = "Category name is required")
        private String category;

        private String description;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class UpdateTransactionRequest {
        @Positive(message = "Amount must be a positive number")
        private BigDecimal amount;

        private String category;

        private String description;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class TransactionResponse {
        private Long id;
        private BigDecimal amount;
        private LocalDate date;
        private String category;
        private String description;
        private CategoryType type;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class TransactionListResponse {
        private List<TransactionResponse> transactions;
    }
}

File: src/main/java/com/syfe/personalfinance/dto/TransactionDto.kt
package com.syfe.personalfinance.dto

import com.syfe.personalfinance.enums.CategoryType
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.PastOrPresent
import jakarta.validation.constraints.Positive
import java.math.BigDecimal
import java.time.LocalDate

class TransactionDto {

    data class CreateTransactionRequest(
        @field:NotNull(message = "Transaction amount is required")
        @field:Positive(message = "Amount must be a positive number")
        val amount: BigDecimal? = null,

        @field:NotNull(message = "Transaction date is required")
        @field:PastOrPresent(message = "Transaction date cannot be in the future")
        val date: LocalDate? = null,

        @field:NotBlank(message = "Category name is required")
        val category: String = "",

        val description: String? = null
    )

    data class UpdateTransactionRequest(
        @field:Positive(message = "Amount must be a positive number")
        val amount: BigDecimal? = null,

        val category: String? = null,

        val description: String? = null
    )

    data class TransactionResponse(
        val id: Long? = null,
        val amount: BigDecimal? = null,
        val date: LocalDate? = null,
        val category: String? = null,
        val description: String? = null,
        val type: CategoryType? = null
    )

    data class TransactionListResponse(
        val transactions: List<TransactionResponse> = emptyList()
    )
}
